UI/CombatReport/CombatLogWnd.cpp

```diff
@@ -66,52 +66,145 @@ class CombatLogWnd::Impl {
 };
 
 namespace {
-    typedef std::shared_ptr<LinkText> LinkTextPtr;
-
-    const std::string EMPTY_STRING;
-
-    // Counts how many ships or populated planets in \a objects are owned by
-    // each empire. Assumes ownership and populations now are consistent with
-    // whenever the result is supposed to indicate, not accounting for 
-    std::map<int, int> CountByOwner(const std::set<int>& owners, const std::set<int>& objects) {
-        std::map<int, int> objects_per_owner;
-        for (int owner_id : owners)
-            objects_per_owner[owner_id] = 0;
-        for (int obj_id : objects) {
-            auto object = Objects().Object(obj_id); // gets destroyed objects, so this function can be used for initial combat forces even if some are destroyed during the combat
-            if (object && (
-                    object->ObjectType() == OBJ_SHIP || (
-                        object->GetMeter(METER_POPULATION) &&
-                        object->InitialMeterValue(METER_POPULATION) > 0.0f)))
-            {
-                int owner_id = object->Owner();
-                if (!objects_per_owner.count(owner_id))
-                    objects_per_owner[owner_id] = 0;
-                ++objects_per_owner[owner_id];
+
+    // TODO: Function adapted from CombatEvents.cpp, will need to be extracted to a common library
+    const std::string& LinkTag(UniverseObjectType obj_type) {
+        static const std::string EMPTY_STRING("");
+
+        switch (obj_type) {
+        case OBJ_SHIP:
+            return VarText::SHIP_ID_TAG;
+        case OBJ_FLEET:
+            return VarText::FLEET_ID_TAG;
+        case OBJ_PLANET:
+            return VarText::PLANET_ID_TAG;
+        case OBJ_BUILDING:
+            return VarText::BUILDING_ID_TAG;
+        case OBJ_SYSTEM:
+            return VarText::SYSTEM_ID_TAG;
+        case OBJ_FIELD:
+        case OBJ_FIGHTER:
+        default:
+            return EMPTY_STRING;
+        }
+    }
+
+    // TODO: Function adapted from CombatEvents.cpp, will need to beextracted to a common library
+    std::string WrapWithTagAndId(const std::string& meat, const std::string& tag, int id) {
+        return boost::str(boost::format("<%1% %2%>%3%</%1%>") % tag % id % meat);
+    }
+
+    /// Returns true, if the \a object should be counted in initial and/or destroyed empire forces
+    bool ShouldObjectBeShownInEmpireForces(std::shared_ptr<UniverseObject> object) {
+        return object && (object->ObjectType() == OBJ_SHIP ||
+                          (object->GetMeter(METER_POPULATION) &&
+                           object->InitialMeterValue(METER_POPULATION) > 0.0f));
+    }
+
+    using EmpireForces = std::vector<std::shared_ptr<UniverseObject>>;
+
+    /// Selects ships and populated planets in \a objects and stores them in collections
+    /// for each empire.
+    std::map<int, EmpireForces> ForcesByOwner(const std::set<int>& owners,
```

Comment:
The `owners` parameter appears to be unused.

Reply:
Yes. It was there already when I started refactoring. I should be paying more attention.